### PR TITLE
feat/improve kucoin perp trade update logic

### PR DESCRIPTION
**Pull request by ethicrypt**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
The logic of handling websocket updates is improved based on feedback from @cardosofede and is now more in line with how the Kucoin spot connector operates. 

**Tests performed by the developer**:
I ran all of the unit tests, as well as running a strategy on the testnet and a strategy on the mainnet. Everything worked as expected.

**Tips for QA testing**:
The amount of testing that can be done by running a strategy on testnet is minimal, because orders are virtually never filled.

### CODE REVIEW

**Comment by rapcmia** [comment]
PR update:
While testing trade filled events we observed that short position is opened however it did not create a buy take-profit instead it created a sell order:
- The client is able to cancel the other side of the orderbook in favor of take-order
- The client created LIMIT order but did not indicate which side {buy, side}
- Then getting `unexpected error in user stream error`
    ```
    2023-03-23 16:22:01,154 - 7899 - hummingbot.connector.derivative.kucoin_perpetual.kucoin_perpetual_derivative.KucoinPerpetualDerivative - ERROR - Unexpected error in user stream listener loop.
    Traceback (most recent call last):
      File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 518, in _user_stream_event_listener
        self._process_trade_event_message(execution_data)
      File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 576, in _process_trade_event_message
        trade_update = self._parse_trade_update(trade_msg=trade_msg, tracked_order=fillable_order)
      File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 618, in _parse_trade_update
        exec_price = Decimal(trade_msg["price"])
    KeyError: 'price'
    ```
- The client then created a sell instead of buy order to close the position
  - Sell order filled
  - Checked exchange, the order was labeled as buy
    image
    image

This has been observed on kucoin_perpetual on development branch

[logs_perp-kucoin1.log](https://github.com/hummingbot/hummingbot/files/11048727/logs_perp-kucoin1.log)
[trades_perp-kucoin1.csv](https://github.com/hummingbot/hummingbot/files/11048737/trades_perp-kucoin1.csv)

**Comment by ethicrypt** [comment]
> PR update: While testing trade filled events we observed that short position is opened however it did not create a buy take-profit instead it created a sell order:
> 
> * The client is able to cancel the other side of the orderbook in favor of take-order
> * The client created LIMIT order but did not indicate which side {buy, side}
> * Then getting `unexpected error in user stream error`
>   ```
>   2023-03-23 16:22:01,154 - 7899 - hummingbot.connector.derivative.kucoin_perpetual.kucoin_perpetual_derivative.KucoinPerpetualDerivative - ERROR - Unexpected error in user stream listener loop.
>   Traceback (most recent call last):
>     File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 518, in _user_stream_event_listener
>       self._process_trade_event_message(execution_data)
>     File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 576, in _process_trade_event_message
>       trade_update = self._parse_trade_update(trade_msg=trade_msg, tracked_order=fillable_order)
>     File "/Users/ralphcomia/hbot/github/hummingbot/6161/hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py", line 618, in _parse_trade_update
>       exec_price = Decimal(trade_msg["price"])
>   KeyError: 'price'
>   ```
> * The client then created a sell instead of buy order to close the position
>   
>   * Sell order filled
>   * Checked exchange, the order was labeled as buy
>     image
>     image
> 
> This has been observed on kucoin_perpetual on development branch
> 
> [logs_perp-kucoin1.log](https://github.com/hummingbot/hummingbot/files/11048727/logs_perp-kucoin1.log) [trades_perp-kucoin1.csv](https://github.com/hummingbot/hummingbot/files/11048737/trades_perp-kucoin1.csv)

@rapcmia Would you provide the strategy file you used for testing? I think it would be helpful for me to debug this under the same conditions.

**Comment by rapcmia** [comment]
hi @ethicrypt apologies I forgot to add it earlier, please attached:
[perp-kucoin1.yml.zip](https://github.com/hummingbot/hummingbot/files/11052349/perp-kucoin1.yml.zip)

Adding to the results earlier, what we did is to play with the long/short taking spread using 0.5 and 1.0. Hope this helps

File: hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_constants.py [hunk 1]
```diff
@@ -89,6 +89,11 @@
 WS_CONNECTION_TIME_INTERVAL = 20
 WS_REQUEST_LIMIT_ID = "WSRequest"
 
+ORDER_CHANGE_EVENT_TYPE = "orderChange"
+BALANCE_EVENT_TYPE = "availableBalance.change"
+POSITION_CHANGE_EVENT_TYPE = "position.change"
+TRADE_EVENT_TYPE = "match"
```
Comment (cardosofede) [inline hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_constants.py:95]: this should be added to the dictionary of OrderState

File: hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py [hunk 2]
```diff
@@ -506,32 +506,24 @@ async def _user_stream_event_listener(self):
...
+                execution_data = event_message.get("data")
                 endpoint = web_utils.endpoint_from_message(event_message)
                 payload = web_utils.payload_from_message(event_message)
-
-                if endpoint == CONSTANTS.WS_SUBSCRIPTION_POSITIONS_ENDPOINT_NAME:
+                if type(execution_data) == list:
+                    execution_data = execution_data[0]
+
+                if event_type == "message" and event_subject == CONSTANTS.ORDER_CHANGE_EVENT_TYPE:
+                    self._process_trade_event_message(execution_data)
```
Comment (cardosofede) [inline hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py:518]: Nice refactor, now is needed to improve the method _process_trade_event_message. Check the logic implemented in Kucoin exchange, is more clean. Also the bug that Ralph is referring is addressed there. the name is not price is MatchPrice or something like that

**Review by cardosofede — CHANGES_REQUESTED** [review]

File: hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_constants.py (hunk 1 above)
Comment (ethicrypt) [reply hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_constants.py:95]: I'm a bit confused as to why these would be added to the OrderState dictionary. These aren't order states, they are the types of WS events being returned. This is a separate piece of data from the state of the order, as I understand it. Apologies if I'm off base, just want to be sure this is done right.

**Comment by ethicrypt** [comment]
I have added new changes to address the failing unit tests and improve the trade/order update logic.

File: hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py (hunk 2 above)
Comment (ethicrypt) [reply hummingbot/connector/derivative/kucoin_perpetual/kucoin_perpetual_derivative.py]: There have been further improvements here in my latest check-in, do these address your concerns?

**Review by cardosofede — DISMISSED** [review]
LGTM!

**Comment by ethicrypt** [comment]
An issue was identified with the recorded fees with Kucoin Perpetual, I'm currently investigating that and will provide fix as part of this PR.

**Comment by ethicrypt** [comment]
All known issues should be resolved now.

**Comment by nikspz** [comment]
@ethicrypt There's a bug on current order to CLOSE behavior latest commit [f41d7d7](https://github.com/hummingbot/hummingbot/pull/6161/commits/f41d7d785cc65f0b116c5674204726c68b63a4d5)

Steps to reproduce:
Clone feature branch and start instance
Create perpetualMM using kucoin perpetual
Set up take profit for long and short to 0.3 and stop loss to 0.2
Start the strategy and wait till order to OPEN filled
Pay attention to bot behavior on creation orders to CLOSE

Actual behavior:
Bot created and filled SELL order to OPEN
Bot spamed with creating stop loss sell order
Order not closed with take profit BUY order to CLOSE

Expected:
Bot will create take profit BUY order to CLOSE using short_profit_taking_spread
after CLOSE order filled bot will create orders to OPEN



[6161kucoinperpfix-commitf41d.zip](https://github.com/hummingbot/hummingbot/files/11072117/6161kucoinperpfix-commitf41d.zip)

image
image

**Comment by ethicrypt** [comment]
> @ethicrypt There's a bug on current order to CLOSE behavior latest commit [f41d7d7](https://github.com/hummingbot/hummingbot/pull/6161/commits/f41d7d785cc65f0b116c5674204726c68b63a4d5)
> 
> Steps to reproduce:
> 
> 1. Clone feature branch and start instance
> 2. Create perpetualMM using kucoin perpetual
> 3. Set up take profit for long and short to 0.3 and stop loss to 0.2
> 4. Start the strategy and wait till order to OPEN filled
> 5. Pay attention to bot behavior on creation orders to CLOSE
> 
> Actual behavior: Bot created and filled SELL order to OPEN Bot spamed with creating stop loss sell order Order not closed with take profit BUY order to CLOSE at 2.42% (should be at 0.3)
> 
> Expected: Bot will create take profit BUY order to CLOSE using short_profit_taking_spread after CLOSE order filled bot will create orders to OPEN
> 
> [6161kucoinperpfix-commitf41d.zip](https://github.com/hummingbot/hummingbot/files/11072117/6161kucoinperpfix-commitf41d.zip)
> 
> image image

I followed the steps to reproduce and actually did not run into any issues like this. Based on the logs, it seems like it is running into rate limit issues and that's the true source of the issue... But it doesn't seem like the Kucoin Perp limits are wrong. I tried lowering the limits and that seemed to help, but it seems there is something deeper here. Going to keep investigating.

**Comment by rapcmia** [comment]
While checking long term bot, tried to close and relaunch to check position retrieve behavior and we noticed that it tries to create and eventually cancel an order. Observed that the unrealized PnL is incorrect and assuming this is causing the bot recognize it and starts creating orders

image

Left the bot for couple of minutes and this happens, a lot of create/cancel order at a span of seconds for every 60s
image

Steps to reproduce
1. Setup a perp-mm
2. Observed behavior for position retrieval for both long and short position

[6161.zip](https://github.com/hummingbot/hummingbot/files/11076921/6161.zip)

**Comment by ethicrypt** [comment]
The latest commit should address the unrealized PnL issues, based on my testing.

**Comment by cardosofede** [comment]
@ethicrypt you will need to add a test for the last commit since the coverage now is 77%

**Comment by rapcmia** [comment]
> The latest commit should address the unrealized PnL issues, based on my testing.

Thanks for the update, while checking this again it is observed that the issue still persist. The unrealized Pnl still inconsistent resulting to unexpected order create and cancel order at same time 
image

It is observed that the unrealized Pnl is positive yet on the client it is negative
image

Steps to reproduce:
1. Open a short position and observe connector behavior 

[6161.zip](https://github.com/hummingbot/hummingbot/files/11106693/6161.zip)

**Pull request closed without merging**